doc/developer/design/20210412_postgres_sources.md

```diff
@@ -0,0 +1,210 @@
+  implementation. Future optimization could lift this constraint.
+
+* **no columns with TOASTed text**. TOAST is a way for postgres to store large
+  amounts of text. However, these values are not present in the replication
+  stream even when `REPLICA IDENTITY` is set to `FULL`. If a TOASTed value is
+  encountered the source will enter an errored state.
+
+* **The TCP connection to the upstream database will not drop**. Resuming a
+  replication slot is a bit complicated so the MVP will assume the initial
+  connection will remain open. Lifting this assumption should be prioritized
+  after the MVP.
+
+
+## Description
+
+### Terminology
+
+Logical replication involves a few different postgres objects that work together
+to move data from one postgres instance to another.
+
+1. **Logical replication**: This is a special mode of postgres streaming
+   replication where data is streamed transaction by transaction in commit
+   order. This is in contrast with normal replication where binary segments of
+   the WAL are sent to the downstream database.
+2. **Decoding plugin**: An component that runs inside the upstream database that
+   is responsible for reading the raw WAL segments and reconstructing the
+   logical transaction messages before sending them over to the logical
+   replication receiver.
+3. **pgoutput**: The specific decoder that postgres uses when logical
+   replication is used. There are a lot of third-party plugins available but
+   pgoutput is offered by postgres itself.
+4. **PUBLICATION**: It is an object that tells the upstream database which
+   tables are to be published over a replication stream. This can be an explicit
+   list of table names or a special "all tables" publication that gets
+   everything.
+5. **replication slot**: It is an object that keeps track of how far along a
+   replica is with its stream consumption. Normally the upstream database has
+   one such object for each downstream replica. Each replica periodically sends
+   updates with the WAL offset up to which it has consumed which then lets the
+   upstream database safely clean up segments from disk.
+
+### Replication mechanism
+
+In order to replicate a set of tables from an upstream database a replica needs
+to orchestrate two phases, making sure that there is no data loss during the
+handover. The first phase involves the initial table sync which is done with
+normal `COPY` bulk transfer commands. Once the replica has the initial snapshot
+of a table it can switch to reading updates from a replication slot in order to
+keep its table synced.
+
+Ensuring that there is no gap between finishing the initial COPY and starting
+reading from the replication slot is done by synchronizing the database snapshot
+at which the `COPY` is performed with the starting point of the replication
+slot. The simplest way to do this is by using the following query structure:
+
+```sql
+-- Initial sync phase
+BEGIN READ ONLY ISOLATION LEVEL REPEATABLE READ
```

Comment:
What impact does this have on the database performance? Can this be run on a primary database, or should we recommend that people replicate tables from a replica database?

Reply:
Yes, but you would only have to provision the Pg read replica once, and then if you are bringing Materialize up and down multiple times (or it crashes) that load would go to the replica.

Reply:
That's a good point. So I think our documentation should say that every time materialize starts up it's as if a new postgres replica is connected to the upstream database, and let the db administrators decide if this an acceptable load for their primary